viewing_party/party.py

```diff
@@ -1,23 +1,153 @@
+
+    return total / len(watched)
+
+def get_most_watched_genre(user_data):
+    # If the user hasn't watched any movies, return None
+    if not user_data["watched"]:
+        return None
+
+    genre_count = {}
+
+    # Loop through watched movies and get genre of current movie
+    for movie in user_data["watched"]:
+        genre = movie["genre"] 
+        
+        if genre in genre_count:
+            genre_count[genre] += 1
+
+        else:
+            genre_count[genre] = 1
+
+    # return genre with highest count of movies
+    most_watched = max(genre_count, key=genre_count.get)
+    return most_watched
+
+    
+# get list of movies user has watched, but friends haven't.
+def get_unique_watched(user_data):
+    unique_movies = []
+
+    friends_watched = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in friends_watched:
+                friends_watched.append(movie)
 
-# -----------------------------------------
-# ------------- WAVE 2 --------------------
-# -----------------------------------------
+    for movie in user_data["watched"]:
+        if movie not in friends_watched:
+            unique_movies.append(movie)
 
+    return unique_movies
 
-# -----------------------------------------
-# ------------- WAVE 3 --------------------
-# -----------------------------------------
 
+def get_friends_unique_watched(user_data):
+    friends_movie_for_user = []
+
+    user_watched = set()
+    for movie in user_data["watched"]:
+        user_watched.add(movie["title"])
+
+    friends_watched = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched and movie["title"] not in friends_watched:
+                friends_movie_for_user.append(movie)
+                friends_watched.add(movie["title"])
+
+    return friends_movie_for_user
+        
         
-# -----------------------------------------
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
+def get_available_recs(user_data):
+    recommended_list = []
+    subscriptions = set(user_data.get("subscriptions", []))
+    user_watched_titles = {movie["title"] for movie in user_data.get("watched", [])}
+    recommended_titles = set()
+
+    for friend in user_data.get("friends", []):
+        for movie in friend.get("watched", []):
+            if (
+                movie["title"] not in user_watched_titles
+                and movie["host"] in subscriptions
+                and movie["title"] not in recommended_titles
+            ):
+                recommended_list.append(movie)
+                recommended_titles.add(movie["title"])
+
+    return recommended_list
+
+def get_new_rec_by_genre(user_data):
+    recommendations = []
+
+    # Get most watched genre
+    most_genre = get_most_watched_genre(user_data)
+
+    if most_genre is None:
+        return recommendations

```

Comment:
We declare the `recommendations` and `most_genre` variables before we know if the data is valid and those variables will be used. We could move our guard clause up some by checking if our input list itself is empty and only create those variables if they will be needed:

```py
if (not user_data) or (not user_data['watched']) :
    return []
```